Source/HTTP/iOS/http_ios.mm

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Microsoft Corporation
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+#include "pch.h"
+#include "http_ios.h"
+
+ios_http::ios_http(_In_ AsyncBlock* asyncBlock, _In_ hc_call_handle_t call) :
+    m_call(call),
+    m_asyncBlock(asyncBlock)
+{
+    NSURLSessionConfiguration* configuration = NSURLSessionConfiguration.ephemeralSessionConfiguration;
+
+    uint32_t timeoutInSeconds = 0;
+    if (FAILED(HCHttpCallRequestGetTimeout(m_call, &timeoutInSeconds)))
+    {
+        // default to 60 to match other default ios behaviour
+        timeoutInSeconds = 60;
+    }
+    
+    [configuration setTimeoutIntervalForRequest:(NSTimeInterval)timeoutInSeconds];
+    [configuration setTimeoutIntervalForResource:(NSTimeInterval)timeoutInSeconds];
+    
+    m_session = [NSURLSession sessionWithConfiguration:configuration];
+}
+
+void ios_http::completion_handler(NSData *data, NSURLResponse *response, NSError *error)
+{
+    ASSERT([response isKindOfClass:[NSHTTPURLResponse class]]);
+    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse*)response;
+    
+    if (error)
+    {
+        // TODO: handle errors
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return;
+    }
+
+    uint32_t statusCode = (uint32_t)[httpResponse statusCode];
+    
+    HCHttpCallResponseSetStatusCode(m_call, statusCode);
+    
+    NSDictionary *headers = [httpResponse allHeaderFields];
+    for (NSString *key in headers)
+    {
+        NSString *value = headers[key];
+        
+        const char *keyCString = [key cStringUsingEncoding:NSUTF8StringEncoding];
+        const char *valueCString = [value cStringUsingEncoding:NSUTF8StringEncoding];
+        HCHttpCallResponseSetHeader(m_call, keyCString, valueCString);
+    }
+
+    HCHttpCallResponseSetResponseBodyBytes(m_call, static_cast<const uint8_t*>([data bytes]), [data length]);
+    CompleteAsync(m_asyncBlock, S_OK, 0);
+}
+
+void ios_http::initiate_request()
+{
+    const char* urlCString = nullptr;
+    const char* methodCString = nullptr;
+    if (FAILED(HCHttpCallRequestGetUrl(m_call, &methodCString, &urlCString)))
+    {
+        CompleteAsync(m_asyncBlock, E_FAIL, 0);
+        return; //TODO: handle errors better
+    }
+    
+    NSString* urlString = [[NSString alloc] initWithUTF8String:urlCString];
```

Comment:
can you add a comment indicating these objects are magically refcounted?

Reply:
I disagree. Auto refcounting is so old in Objective-C at this point that anything else would be unexpected. I don't want to mark that every time we instantiate something. Now, it is fair to ask if we want to file a task to look into switching to custom allocators...

---
In reply to: [188699186](https://github.com/Microsoft/libHttpClient/pull/162#discussion_r188699186) [](ancestors = 188699186)

Reply:
Fair. I just want to be able to distinguish things that we have to manually manage the lifetime for and things that are automatically managed.

Reply:
Agreed offline. Let's mark the c++ memory management in a way that I call out at the end of this file by using smart pointers for c++ memory.

---
In reply to: [188796315](https://github.com/Microsoft/libHttpClient/pull/162#discussion_r188796315) [](ancestors = 188796315)